packages/protocol/contracts/team/proving/ProverSet.sol

```diff
@@ -1,63 +1,92 @@
 // SPDX-License-Identifier: MIT
 pragma solidity 0.8.24;
 
+import "@openzeppelin/contracts-upgradeable/token/ERC20/extensions/ERC20VotesUpgradeable.sol";
 import "@openzeppelin/contracts/interfaces/IERC1271.sol";
 import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "../../common/EssentialContract.sol";
 import "../../common/LibStrings.sol";
 import "../../L1/ITaikoL1.sol";
 
+interface IHasRecipient {
+    function recipient() external view returns (address);
+}
+
 /// @title ProverSet
 /// @notice A contract that holds TKO token and acts as a Taiko prover. This contract will simply
 /// relay `proveBlock` calls to TaikoL1 so msg.sender doesn't need to hold any TKO.
 /// @custom:security-contact security@taiko.xyz
 contract ProverSet is EssentialContract, IERC1271 {
     bytes4 private constant _EIP1271_MAGICVALUE = 0x1626ba7e;
 
-    mapping(address prover => bool isProver) public isProver;
-    uint256[49] private __gap;
+    mapping(address prover => bool isProver) public isProver; // slot 1
+    address public admin; // slot 2
+
+    uint256[48] private __gap;
 
     event ProverEnabled(address indexed prover, bool indexed enabled);
     event BlockProvenBy(address indexed prover, uint64 indexed blockId);
 
     error INVALID_STATUS();
     error PERMISSION_DENIED();
 
+    modifier onlyAuthorized() {
+        if (msg.sender != admin && msg.sender != IHasRecipient(admin).recipient()) {
+            revert PERMISSION_DENIED();
+        }
+        _;
+    }
+
+    modifier onlyProver() {
+        if (!isProver[msg.sender]) revert PERMISSION_DENIED();
+        _;
+    }
+
     /// @notice Initializes the contract.
-    function init(address _owner, address _addressManager) external initializer {
+    function init(
+        address _owner,
+        address _admin,
+        address _addressManager
+    )
+        external
+        nonZeroAddr(_admin)
+        initializer
+    {
         __Essential_init(_owner, _addressManager);
 
-        IERC20 tko = IERC20(resolve(LibStrings.B_TAIKO_TOKEN, false));
+        admin = _admin;
 
-        address taikoL1 = resolve(LibStrings.B_TAIKO, false);
-        tko.approve(taikoL1, type(uint256).max);
-
-        address assignmentHook = resolve(LibStrings.B_ASSIGNMENT_HOOK, false);
-        tko.approve(assignmentHook, type(uint256).max);
+        IERC20 tko = IERC20(resolve(LibStrings.B_TAIKO_TOKEN, false));
+        tko.approve(resolve(LibStrings.B_TAIKO, false), type(uint256).max);
+        tko.approve(resolve(LibStrings.B_ASSIGNMENT_HOOK, false), type(uint256).max);
     }
 
     /// @notice Enables or disables a prover.
-    function enableProver(address _prover, bool _isProver) external onlyOwner {
+    function enableProver(address _prover, bool _isProver) external onlyAuthorized {
```

Comment:
I guess we need to put this function in the `TokenUnlock` too, if we want to allow unlocking people calling it.

Reply:
why the unlocking needs to know about prover enabling...

Reply:
Yes, okay i see the modifier is like that:
`msg.sender != IHasRecipient(admin).recipient()`